prow/bugzilla/client.go

```diff
@@ -214,37 +231,112 @@ func getRootForClone(c Client, bug *Bug) (*Bug, error) {
 
 // GetRootForClone returns the original bug.
 func (c *client) GetRootForClone(bug *Bug) (*Bug, error) {
-	return getRootForClone(c, bug)
+	bugCache := map[int]*Bug{}
+	return getRootForClone(c, bug, bugCache)
 }
 
 // GetAllClones returns all the clones of the bug including itself
 // Differs from GetClones as GetClones only gets the child clones which are one level lower
-func (c *client) GetAllClones(bug *Bug) ([]*Bug, error) {
-	return getAllClones(c, bug)
+func (c *client) GetAllClones(bugID int) ([]*Bug, *Bug, *Bug, error) {
+	bugCache := map[int]*Bug{}
+	return getAllClones(c, bugID, bugCache)
 }
-func getAllClones(c Client, bug *Bug) ([]*Bug, error) {
-	root, err := getRootForClone(c, bug)
+
+type clonesCacheWithMux struct {
+	cache map[int][]int
+	mux   sync.Mutex
+}
+
+var clonesCache = clonesCacheWithMux{cache: map[int][]int{}}
+
+func getAllClones(c Client, bugID int, intraCallCache map[int]*Bug) ([]*Bug, *Bug, *Bug, error) {
+
+	// Global clones cache should only be accessed after ensuring proper thread safety
+	clonesCache.mux.Lock()
+	clonesID, ok := clonesCache.cache[bugID]
+	clonesCache.mux.Unlock()
+	var bug *Bug
+	var getBugErr error
+	var wg sync.WaitGroup
+	cloneErrors := make([]error, len(clonesID))
+	clones := make([]*Bug, len(clonesID))
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// This error is being handled after the wg.Wait
+		bug, getBugErr = c.GetBug(bugID)
+	}()
+	if ok {
+		var intraCallCacheMux sync.Mutex
+		for index, cloneID := range clonesID {
+			wg.Add(1)
+			go func(index int, cloneID int) {
+				defer wg.Done()
+				var clone *Bug
+				intraCallCacheMux.Lock()
+				clone, ok = intraCallCache[cloneID]
+				intraCallCacheMux.Unlock()
+				if !ok {
+					var err error
+					clone, err = c.GetBug(cloneID)
+					if err != nil {
+						cloneErrors[index] = err
+						return
+					}
+				}
+				clones[index] = clone
+			}(index, cloneID)
+		}
+	}
+	wg.Wait()
+	intraCallCache[bug.ID] = bug
```

Comment:
This is super confusing to read, why do we assign here rather than where we retrieved the bug?

Reply:
Assigning it here to keep it threadsafe, I would have to guard it if I were to assign it earlier.

Reply:
Yeah but whats wrong with guarding it? Just abstract that away via a Setter as explained above. The lock is only needed for accessing the cache which is super quick so I doubt this will cause any kind of lock contention